SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -21,62 +21,113 @@ SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const PhysicalWorkerName&
-void SpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentAdded(const Worker_AddComponentOp& Op)
 {
-	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
-		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+	if (!(Op.data.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID || Op.data.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID))
 	{
-		const SpatialGDK::AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(AuthOp.entity_id);
-		if (AuthorityIntentComponent == nullptr)
-		{
-			// TODO(zoning): There are still some entities being created without an authority intent component.
-			// For example, the Unreal created worker entities don't have one. Even though those won't be able to
-			// transition, we should have the intent component on them for completeness.
-		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), AuthOp.entity_id);
-			return;
-		}
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
+			TEXT("Load balancer notified of add component %d for entity %lld which is not a load balancing component"),
+			Op.data.component_id, Op.entity_id);
+		return;
+	}
 
-		const PhysicalWorkerName* OwningWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
-		if (OwningWorkerId != nullptr &&
-			*OwningWorkerId == WorkerId &&
-			StaticComponentView->HasAuthority(AuthOp.entity_id, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
-		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("No need to queue newly authoritative entity %lld because this worker is already authoritative."), AuthOp.entity_id);
-			return;
-		}
-		QueueAclAssignmentRequest(AuthOp.entity_id);
+	if (CanEnforce(Op.entity_id))
+	{
+		MaybeQueueAclAssignmentRequest(Op.entity_id);
+	}
+}
+
+void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentRemoved(const Worker_RemoveComponentOp& Op)
+{
+	if (!(Op.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID || Op.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
+			TEXT("Load balancer notified of remove component %d for entity %lld which is not a load balancing component"),
+			Op.component_id, Op.entity_id);
+		return;
+	}
+
+	if (AclAssignmentRequestIsQueued(Op.entity_id))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
+			TEXT("Component %d for entity %lld removed. Can no longer enforce the previous request for this entity."),
+			Op.component_id, Op.entity_id);
+		return;
+	}
+}
+
+void SpatialLoadBalanceEnforcer::OnEntityRemoved(const Worker_RemoveEntityOp& Op)
+{
+	if (AclAssignmentRequestIsQueued(Op.entity_id))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
```

Comment:
Should the request be dequeued? Here and in OnLoadBalancingComponentRemoved?

Reply:
Also, is this a warning, could it just be log/verbose? Seems like something we should ultimately expect to happen and support - the worker gaining authority should correctly make the assignment when it picks up authority?

Reply:
I think it should be dequeued, however that is a logical change- the previous check for this in processQueued... had a comment saying "not sure if this is possible" and included a warning. This checks for the same thing, so matched the log level. Happy to dequeue and make it a normal log instead.

Reply:
done